File: spin-rust/container/api/src/routes/teams.rs
use super::{ApiError, AppState};
use crate::routes::mappers::{team_attributes_from_row, team_from_row};
use axum::Json;
use axum::extract::{Path, State};
use axum::http::StatusCode;
use axum::response::{IntoResponse, Response};
use football_shared::domain::teams::TeamRecord;

pub(crate) async fn get_team_by_id(
    State(state): State<AppState>,
    Path(id): Path<i32>,
) -> Result<Response, ApiError> {
    let row = sqlx::query("SELECT * FROM team WHERE id = $1")
        .bind(id)
        .fetch_optional(&state.pool)
        .await?;

    match row {
        None => Ok(StatusCode::NOT_FOUND.into_response()),
        Some(row) => Ok(Json(team_from_row(&row)?).into_response()),
    }
}

pub(crate) async fn get_team_by_api_id(
    State(state): State<AppState>,
    Path(id): Path<i32>,
) -> Result<Response, ApiError> {
    let row = sqlx::query("SELECT * FROM team WHERE team_api_id = $1")
        .bind(id)
        .fetch_optional(&state.pool)
        .await?;

    match row {
        None => Ok(StatusCode::NOT_FOUND.into_response()),
        Some(row) => Ok(Json(team_from_row(&row)?).into_response()),
    }
}

pub(crate) async fn get_team_record_by_id(
    State(state): State<AppState>,
    Path(id): Path<i32>,
) -> Result<Response, ApiError> {
    let row = sqlx::query("SELECT * FROM team WHERE team_api_id = $1")
        .bind(id)
        .fetch_optional(&state.pool)
        .await?;

    let team = match row {
        None => return Ok(StatusCode::NOT_FOUND.into_response()),
        Some(r) => team_from_row(&r)?,
    };

    let rows = sqlx::query(
        "SELECT * FROM team_attributes WHERE team_api_id = $1 AND team_fifa_api_id = $2",
    )
    .bind(team.team_api_id)
    .bind(team.team_fifa_api_id)
    .fetch_all(&state.pool)
    .await?;

    let mut attributes = Vec::with_capacity(rows.len());
    for row in rows {
        attributes.push(team_attributes_from_row(&row)?);
    }

    Ok(Json(TeamRecord { team, attributes }).into_response())
}

File: spin-js/src/routes/teams.js
import { jsonResponse, notFound } from '../http.js';
import { teamAttributesFromRow, teamFromRow } from '../mappers.js';

export function registerTeamRoutes(router, { getConnection, parseId }) {
    router.get('/teams/api-id/:id', async ({ id }) => {
        try {
            const connection = getConnection();
            const rowSet = await connection.query('SELECT * FROM team WHERE team_api_id = $1', [parseId(id)]);

            if (!rowSet.rows.length) {
                return notFound();
            }

            return jsonResponse(teamFromRow(rowSet.rows[0]));
        } catch (error) {
            if (error.status === 400) {
                return jsonResponse({ status: 400, error: error.message }, 400);
            }
            throw error;
        }
    });

    router.get('/teams/record/:id', async ({ id }) => {
        try {
            const connection = getConnection();
            const teamRowSet = await connection.query('SELECT * FROM team WHERE id = $1', [parseId(id)]);

            if (!teamRowSet.rows.length) {
                return notFound();
            }

            const team = teamFromRow(teamRowSet.rows[0]);
            const attributesRowSet = await connection.query(
                'SELECT * FROM team_attributes WHERE team_api_id = $1 AND team_fifa_api_id = $2',
                [team.apiId, team.fifaApiId],
            );

            const attributes = attributesRowSet.rows.map(teamAttributesFromRow);
            return jsonResponse({ team, attributes });
        } catch (error) {
            if (error.status === 400) {
                return jsonResponse({ status: 400, error: error.message }, 400);
            }
            throw error;
        }
    });

    router.get('/teams/:id', async ({ id }) => {
        try {
            const connection = getConnection();
            const rowSet = await connection.query('SELECT * FROM team WHERE id = $1', [parseId(id)]);

            if (!rowSet.rows.length) {
                return notFound();
            }

            return jsonResponse(teamFromRow(rowSet.rows[0]));
        } catch (error) {
            if (error.status === 400) {
                return jsonResponse({ status: 400, error: error.message }, 400);
            }
            throw error;
        }
    });
}
